File: number_of_subsequences_that_satisfy_the_given_sum_condition.py
class Solution:
    def numSubseq(self, nums: List[int], target: int) -> int:
        nums.sort()
        left, right = 0, len(nums) - 1
        ans = 0
        mod = 10 ** 9 + 7

        while left <= right:
            if nums[left] + nums[right] > target:
                right -= 1
            else:
                # pow(x, y, z) = x^y % z
                ans += pow(2, right - left, mod)
                left += 1
        
        return ans % mod

File: number_of_subsequences_that_satisfy_the_given_sum_condition.cpp
class Solution {
public:
    int numSubseq(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        int n = nums.size();
        vector<int> pow(n, 1);
        const int mod = 1e9 + 7;
        
        for(int i = 1; i < n; i++){
            pow[i] = (2 * pow[i-1]) % mod;
        }
        
        int left = 0, right = n-1;
        long long res = 0;
        
        while(left <= right){
            if(nums[left] + nums[right] > target){
                right--;
            } else {
                res = (res + pow[right - left]) % mod;
                left++;
            }
        }
        
        return res;
    }
};
